File: app/src/main/java/de/thu/recipebook/models/RecipeListAdapter.java
package de.thu.recipebook.models;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.provider.BaseColumns;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import de.thu.recipebook.R;
import de.thu.recipebook.databases.FavoritesDbHelper;

public class RecipeListAdapter extends BaseAdapter {
    private FavoritesDbHelper dbHelper;
    private List<Recipe> data;

    public RecipeListAdapter(FavoritesDbHelper dbHelper) {
        this.dbHelper = dbHelper;
        data = new ArrayList<>();
    }

    @Override
    public int getCount() {
        return data.size();
    }

    @Override
    public Object getItem(int i) {
        return data.get(i);
    }

    @Override
    public long getItemId(int i) {
        return i;
    }

    @Override
    public View getView(int i, View view, ViewGroup viewGroup) {
        Context context = viewGroup.getContext();
        if (view == null) {
            LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            view = inflater.inflate(R.layout.list_view_item_recipe, null, false);
        }

        ImageView countryImageView = view.findViewById(R.id.image_view_country);
        String country = data.get(i).getCountry();
        String imageResourceName = CountryCollection.getImageResourceName(country);
        int imageId = context.getResources().getIdentifier(imageResourceName,
                "drawable", context.getPackageName());
        countryImageView.setImageResource(imageId);

        TextView textView = view.findViewById(R.id.text_view_recipe_list);
        String name = data.get(i).getName();
        textView.setText(name);

        ImageView favoriteImageView = view.findViewById(R.id.image_view_favorite);
        SQLiteDatabase db = dbHelper.getReadableDatabase();
        Cursor c = db.query(FavoritesDbHelper.FAVORITES_TABLE, new String[]{BaseColumns._ID},
                BaseColumns._ID + " = ?", new String[]{data.get(i).getId()}, null, null, null);
        if (c.getCount() > 0) {
            favoriteImageView.setVisibility(View.VISIBLE);
        } else {
            favoriteImageView.setVisibility(View.INVISIBLE);
        }
        return view;
    }

    public void setData(List<Recipe> data) {
        this.data = data;
    }
}

File: Exercise 8/Kotlin/src/main/java/de/thu/recipebook/models/RecipeListAdapter.kt
package de.thu.recipebook.models

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.provider.BaseColumns
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView
import de.thu.recipebook.R
import de.thu.recipebook.databases.FavoritesDbHelper
import de.thu.recipebook.models.CountryCollection.getImageResourceName

class RecipeListAdapter(private val favoritesDbHelper: FavoritesDbHelper) : BaseAdapter() {
    private var data: List<Recipe>

    init {
        data = ArrayList()
    }

    override fun getCount(): Int {
        return data.size
    }

    override fun getItem(i: Int): Any {
        return data[i]
    }

    override fun getItemId(i: Int): Long {
        return i.toLong()
    }

    override fun getView(i: Int, view: View?, viewGroup: ViewGroup): View {
        var view = view
        val context = viewGroup.context
        if (view == null) {
            val inflater =
                context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            view = inflater.inflate(R.layout.list_view_item_recipe, null, false)
        }

        val country = data[i].country
        val imageResourceName = getImageResourceName(country)
        val imageId = context.resources.getIdentifier(
            imageResourceName,
            "drawable", context.packageName
        )
        view?.findViewById<ImageView>(R.id.image_view_country)?.setImageResource(imageId)

        val textView = view?.findViewById<TextView>(R.id.text_view_recipe_list)
        val name = data[i].name
        textView?.text = name

        val favoriteImageView = view!!.findViewById<ImageView>(R.id.image_view_favorite)
        val db: SQLiteDatabase = favoritesDbHelper.getReadableDatabase()
        val c = db.query(
            FavoritesDbHelper.FAVORITES_TABLE, arrayOf(BaseColumns._ID),
            BaseColumns._ID + " = ?", arrayOf(data[i].id), null, null, null
        )
        if (c.count > 0) {
            favoriteImageView.visibility = View.VISIBLE
        } else {
            favoriteImageView.visibility = View.INVISIBLE
        }


        return view!!
    }

    fun setData(data: List<Recipe>) {
        this.data = data
    }
}